### DISCUSSION TITLE
Increase the number of grains to simulate grain growth

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#16370] [Q&A Modules: Phase field] [answered]
Recently I learned an example of grain growth in the phase field module, "grain_growth_2D_graintracker.i". Based on this input file, I modified the number of grains and order parameters (op_num = 180,grain_num=1000) , increasing the number of grains, but the simulation has problems, as follows：
`Uniformly refining mesh ............                                                       [ 13.40 s]
Caching mesh information                                                                   [  1.65 s]
Initializing equation system .............................................................
...........................................................................                [137.15 s]
Caching mesh information                                                                   [  1.65 s]
Computing max dofs per elem/node .                                                         [  2.53 s]

Framework Information:
MOOSE Version:           git commit ef9631d6e7 on 2020-11-30
LibMesh Version:         0d19e76fc8e7345a0634cba450f488b3a5626eb3
PETSc Version:           3.13.3
SLEPc Version:           3.13.3
Current Time:            Wed Dec  2 09:12:13 2020
Executable Timestamp:    Tue Dec  1 15:13:50 2020

Parallelism:
  Num Processors:          20
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:                   
    Total:                 776161
    Local:                 39202
  Elems:                   
    Total:                 774400
    Local:                 38775
  Num Subdomains:          1
  Num Partitions:          20
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                139708980
  Num Local DOFs:          7056360
  Variables:               { "gr0" "gr1" "gr2" "gr3" "gr4" ... "gr175" "gr176" "gr177" "gr178" "gr179" 
                             } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                3873761
  Num Local DOFs:          194302
  Variables:               "bnds" { "unique_grains" "var_indices" "ghost_regions" "halos" } 
  Finite Element Types:    "LAGRANGE" "MONOMIAL" 
  Approximation Orders:    "FIRST" "CONSTANT" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             IterationAdaptiveDT
  Solver Mode:             Preconditioned JFNK
  PETSc Preconditioner:    hypre boomeramg 

LEGACY MODES ENABLED:
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To remove this message, set 'use_legacy_material
_output' to false in this application. If there are gold output files that contain material property output for which output occurs on INITIAL, then these will generate dif
fs due to zero values being stored, and these tests should be re-golded.

Caching mesh information                                                                   [  1.69 s]
Computing Polycrystal Initial Condition ...........                                        [ 12.19 s]
Projecting initial condition .............................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................                 [2116.46 s]
Computing initial stateful property values ..                                              [  3.86 s]
Initial adaptivity step 1 of 2 ...........................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
....................................................
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 1894809 RUNNING AT pwmoose-PowerEdge-T640
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
`
Any idea what I am doing wrong?
Thanks in advance,
wei peng

**Comment by PengWei97** [comment]
And this is my input file and Display information running in simulation
[pw8.txt](https://github.com/idaholab/moose/files/5628434/pw8.txt)
[grain_growth_2D_graintracker_1000.txt](https://github.com/idaholab/moose/files/5628445/grain_growth_2D_graintracker_1000.txt)

**Comment by laagesen** [comment, marked as answer]
You are probably running out of memory. You don't need to increase the number of order parameters, you should be able to maintain 8 order parameters for a 2D simulation and just increase the number of grains and corresponding mesh size. The advantage of the GrainTracker system is that it will allow you to re-use order parameters. I would suggest reading up on it more here to understand how it works better:

https://mooseframework.inl.gov/source/postprocessors/GrainTracker.html

**Reply by PengWei97** [reply]
> You are probably running out of memory. You don't need to increase the number of order parameters, you should be able to maintain 8 order parameters for a 2D simulation and just increase the number of grains and corresponding mesh size. The advantage of the GrainTracker system is that it will allow you to re-use order parameters. I would suggest reading up on it more here to understand how it works better:
> 
> https://mooseframework.inl.gov/source/postprocessors/GrainTracker.html

Thank you very much for your response. , I got it. 
Now I have another question, in the "voronoi" sub-block, what does "rand_seed = 5" mean? Through the study of PolycrystalVoronoi webpage, I know that it is used to generate a voronoi diagram with 5 subdomains. But if I set "grain_num = 10", the output cloud image is composed of 18 sub-domains and represents 10 grains.
`[UserObjects]
  [./voronoi]
    type = PolycrystalVoronoi
    grain_num = 100 # Number of grains
    rand_seed = 10
  [../]
[]`
So I want to ask：
1. What does "voronoi/rand_seed" affect in the model?
2. And for different numbers of grains, how to choose the appropriate number of "rand_seed"?
Thank you for your precious time. Do you have any information to understand the "voronoi" and "grain_tracker" in the UserObjects block, which is a better Moose program for beginners like me? Learning from the web, I still don't quite understand the meaning of some auxiliary variables in the input file, such as ghost_regions, halos and so on.

**Reply by permcody** [reply]
1. Whenever you see `rand_seed` or something similar, it's a number for seeding the pseudo-random number generator(s) in MOOSE. In this case, each number will give you a different "random" initial condition so you can repeat your simulations as necessary. Feel free to change this number to see what happens to your IC.
2. The rand_seed number has no relation to the number of grains or order parameters. It will affect the location and sizes of all your grains however. Feel free to play around with this number. There is documentation on the Polycrystal Initial Condition system, as well as information on the GrainTracker on this page: [https://mooseframework.inl.gov/modules/phase_field/index.html](). Specifically you can find some guidance on good settings for grains versus order parameters here: [https://mooseframework.inl.gov/source/userobjects/PolycrystalVoronoi.html](). Note however, that if you start to add more and more grains, you will need more and more mesh to capture the smooth interfaces between grains.

**Reply by PengWei97** [reply]
OK，thank you for your help.
